### ISSUE TITLE
Animations Not Importing Properly.

### ISSUE THREAD

**Issue by Mig8806** [niacdoial/blemd#83]
When I import .bmd files, for some reason the animations get messed up. For example when I import Epona from TP her legs kick out in weird directions, as shown in the .gif below. Not sure why this was happening and I was wondering if there is a way to fix it.

![epona walk](https://github.com/niacdoial/blemd/assets/76982162/db1a6715-1680-4607-8091-dfd0d2dd2cd2)

**Comment by spamzilla** [comment]
I'm having the same issue here.

**Comment by Mig8806** [comment]
> I'm having the same issue here. Sometimes the whole front end of the horse will distort as well.

Glad to know I'm not the only one. Let me know if you find a way to fix it.

**Comment by niacdoial** [comment]
oops sorry for the delay
can you let me look at the animation curves for those legs, please?

**Comment by spamzilla** [comment]
Sorry for sounding like a noob but what are 'animation curves' exactly?

**Comment by niacdoial** [comment]
oh no worries. gotta learn at some point, right? (also it wouldn't be the first time I word something weirdly and it makes it impossible to understand)

In short, it's a way to see the different motions in the animation as "amount of X as a function of time"

here are some instructions to reach it (...at least on blender 4.0, tell me if you use another version and I'll try to write new instructions). I'll assume you almost never used blender before, just to be safe.

- the UI of blender is composed of what I call "panes" with distinct roles. by default, there's the 3D view, the timeline, the outliner (the tree-like view on the top-right) and the properties. In the corner of each pane, there's a small icon that allows you to change it to a new type. (want four timelines for some reason? you can!) You can also resize the panes if you want.
- first, you want to go to the top of the blender window, and click on the "animation" tab, which will change what panes are visible.
- then, in either the outliner or the 3D view, select the armature (looks like a stick figure in the outliner, and as diamond-things in the 3D view.)
- then, look for the top-left corner of the timeline-ish pane, and switch that one dropdown-with-text from "Dope Sheet" to "Action editor"
- at the top-center of the pane, a new dropdown-or-create thingy appeared. click the dropdown and select the animation you want to see.
- now, in the 3D view, select the bone (the diamond-like thing) that represents the twitchy leg.
- in the action editor, there should be a bunch of lines that appear, from "X location" to "Z scale". use Ctrl-click to select the ones that say "rotation"
- now, select any pane, and change its type to "graph editor".
- finally, move around and zoom as you please, with the scroll wheel, Ctrl-scroll, Shift-scroll, or by moving the scroll bars (you can move the tips of the scroll bars to zoom or unzoom)
- ~~profit~~

**Comment by spamzilla** [comment]
I think what you're looking for might be something like this? :

![Zelda TP Horse Graph](https://github.com/niacdoial/blemd/assets/48328838/f76f2a52-0fbd-499f-aa4a-f6a333ed4598)

**Comment by niacdoial** [comment]
hm... I might have a few ideas of what's happening.
tell me: does the position of the leg look good at frames 0, 7 and 14?
also, as a hacky workaround, does this problem disappear when, in the select-file-to-import dialog, you enable the "disable axis conversion" option on the right side of the dialog?

**Comment by spamzilla** [comment]
Yes, frames 0, 7, and 14 do appear to be correct. "Disable axis conversion" seems to only rotate the horse by 90 degrees though. The mispositioning of the leg is still present.

**Comment by Mig8806** [comment]
Sorry for the delay. but here is my animation graph:
![image](https://github.com/niacdoial/blemd/assets/76982162/a770f663-b308-40a8-82d0-5eca47519dc7)

**Comment by niacdoial** [comment]
ok, so good news: I think I know exactly why this is happening. (long story short, there are several ways to express the same rotation in an "Euler XZY" representation, and for some reason frame 7 uses a different way than frames 0 and 14)
bad news: I don't know where it happens yet.
also sorry Mig8806 but I can't really see anything on that graph, everything is too cramped together horizontally.

...Also, Can I also ask either of you if the problem exists on blender before 4.0? it seems weird that I get two users with the same bug on the same basic feature after so much time, and so close to each other

**Comment by niacdoial** [comment]
actually, do tell me if the newest commit fixes it (there's a new option on the right side of the choose-file-to-import dialog) that should be enabled for this

**Comment by spamzilla** [comment]
I know this is github but I actually have no clue how git works.

**Comment by niacdoial** [comment]
no problem lol. I'll just... have to guess what part you didn't understand and give an explanation

If it's a matter of not understanding what I meant:<br/>
commits are just... snapshots of the code that has been written. in other words, each of them stores "what the code of this program looked like at that point in time".<br/>
so what I mean by "the newest commit might fix this bug" is "On my own computer, I made some changes to the code which might fix this bug, then took a snapshot of the code, and uploaded that snapshot to github for anyone to download"

in other words, "try downloading BleMD again, the newest version has changes that might fix this."

...hope that helps, if this was your question

**Comment by spamzilla** [comment]
The new option makes things much worse. Everything is rotating like crazy!

**Comment by niacdoial** [comment]
...I see. What about now?

**Comment by niacdoial** [comment]
what about *now*?

**Comment by BeezBumba** [comment]
@Mig8806 Can I ask how you imported the BMD files with animations? I'm trying to do the same thing but like link instead?

**Comment by niacdoial** [comment]
@BeezBumba
the animation files need to be placed in a specific way with respect to the bmd file
(see: https://github.com/niacdoial/blemd?tab=readme-ov-file#how-are-the-the-animations-handled, or more accurately https://github.com/niacdoial/blemd/blob/648205a3c173d5f4c6806d5eba30f20024ba5386/README.md?plain=1#L80 since the markdown rendering is borked right now)

**Comment by BeezBumba** [comment]
When I extract the AlAnm.arc file(has link's animations) it gives me 2 folders that have bmds bmde and bmdr I'm also not sure which bmd is in charge for all of the bcks. ![Screenshot 2024-05-23 161556](https://github.com/niacdoial/blemd/assets/130922034/63b6d2ae-111c-49c8-a27e-f165c124bea3)

**Comment by niacdoial** [comment]
@Mig8806 @spamzilla
I think this issue should be resolved now
sorry I only got the courage to tackle this again after a few months

**Comment by spamzilla** [comment]
Unfortunately it is still totally bugged :-( 

![Blemd](https://github.com/user-attachments/assets/1ee078fa-3d0a-4e02-a556-3d78e511f736)

**Comment by spamzilla** [comment]
I found that Blender 3.5 is far more buggier than Blender 4.0. In 3.5 I can get something like this: 

![Blemd](https://github.com/user-attachments/assets/10e55d50-2625-4c68-8247-2e1c0f224c99)

In version 4.0, most of the contortion is fixed. The buggy parts are only the back legs. They still go out of position a lot of the time. Maybe I should download the newest Blender to verify this?

**Comment by niacdoial** [comment]
goodness!
and that's with the "use 'smallest rotation rule' on animations" checkbox ticked on import?
and yeah I'm not surprised it could change between blender versions, since this bug lies in the fact that one of blender's functions behaves inconsistently between calls

**Comment by spamzilla** [comment]
The "use smallest rotation rule"  feature almost completely fixes the problems

**Comment by niacdoial** [comment]
ah, almost....
What I expect is that this feature can fix animations like this one, but is also liable to break animations with even more ample motion inbetween keyframes, hence why it is not a default.
what part of the error remains when the option is ticked?

**Comment by spamzilla** [comment]
Frames 835-838 appear to be kicking around incorrectly but it is a minor thing

**Comment by niacdoial** [comment]
wasn't that the exact "amplitude" of the bug from when you first contacted me?

**Comment by spamzilla** [comment]
I'm not sure actually...?
